Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp

```diff
@@ -217,59 +229,62 @@ void CoreAudioCaptureSource::initializeToStartProducingData()
+    Ref unit = m_unit;
+    unit->setCaptureDevice(String { persistentID() }, m_captureDeviceID, captureDevice().isDefault());
 
-    bool shouldReconfigure = echoCancellation() != CoreAudioSharedUnit::singleton().enableEchoCancellation() || sampleRate() != CoreAudioSharedUnit::singleton().sampleRate() || volume() != CoreAudioSharedUnit::singleton().volume();
-    CoreAudioSharedUnit::singleton().setEnableEchoCancellation(echoCancellation());
-    CoreAudioSharedUnit::singleton().setSampleRate(sampleRate());
-    CoreAudioSharedUnit::singleton().setVolume(volume());
+    bool shouldReconfigure = echoCancellation() != unit->enableEchoCancellation() || sampleRate() != unit->sampleRate() || volume() != protectedUnit()->volume();
+    unit->setEnableEchoCancellation(echoCancellation());
+    unit->setSampleRate(sampleRate());
+    unit->setVolume(volume());
 
-    CoreAudioSharedUnit::singleton().addClient(*this);
+    unit->addClient(*this);
 
     if (shouldReconfigure)
-        CoreAudioSharedUnit::singleton().reconfigure();
+        protectedUnit()->reconfigure();
 
     m_currentSettings = std::nullopt;
 }
 
 CoreAudioCaptureSource::~CoreAudioCaptureSource()
 {
-    CoreAudioSharedUnit::singleton().removeClient(*this);
+    protectedUnit()->removeClient(*this);
 }
 
 void CoreAudioCaptureSource::startProducingData()
 {
     m_canResumeAfterInterruption = true;
     initializeToStartProducingData();
-    CoreAudioSharedUnit::singleton().startProducingData();
+    protectedUnit()->startProducingData();
     m_currentSettings = { };
 }
 
 void CoreAudioCaptureSource::stopProducingData()
 {
     ALWAYS_LOG_IF(loggerPtr(), LOGIDENTIFIER);
-    CoreAudioSharedUnit::singleton().stopProducingData();
+    protectedUnit()->stopProducingData();
 }
 
 void CoreAudioCaptureSource::endProducingData()
 {
     ALWAYS_LOG_IF(loggerPtr(), LOGIDENTIFIER);
-    CoreAudioSharedUnit::singleton().removeClient(*this);
+
+    Ref unit = m_unit;
+    unit->removeClient(*this);
     if (isProducingData())
-        CoreAudioSharedUnit::singleton().stopProducingData();
+        unit->stopProducingData();
 }
 
 const RealtimeMediaSourceCapabilities& CoreAudioCaptureSource::capabilities()
 {
     if (!m_capabilities) {
-        BaseAudioSharedUnit& unit = CoreAudioSharedUnit::singleton();
+        Ref unit = m_unit.get();
 
         RealtimeMediaSourceCapabilities capabilities(settings().supportedConstraints());
         capabilities.setDeviceId(hashedId());
         capabilities.setGroupId(hashedGroupId());
         capabilities.setEchoCancellation(m_echoCancellationCapability ? (*m_echoCancellationCapability ? RealtimeMediaSourceCapabilities::EchoCancellation::On : RealtimeMediaSourceCapabilities::EchoCancellation::Off) : RealtimeMediaSourceCapabilities::EchoCancellation::OnOrOff);
         capabilities.setVolume({ 0.0, 1.0 });
-        capabilities.setSampleRate(unit.sampleRateCapacities());
+        capabilities.setSampleRate(unit->sampleRateCapacities());
```

Comment:
I guess you could use protectedUnit() here as well. Not sure why it was put in a local variable before.